nbgrader/exchange/exchange.py

```diff
@@ -129,12 +129,31 @@ def do_copy(self, src, dest, log=None):
                                                include=self.coursedir.include,
                                                max_file_size=self.coursedir.max_file_size,
                                                log=self.log))
+        # copytree copies access mode too - so we must add go+rw back to it if
+        # we are in groupshared.
+        if self.coursedir.groupshared:
+            for dirname, _, filenames in os.walk(dest):
+                # dirs become ug+rwx
+                try:
+                    os.chmod(dirname, (os.stat(dirname).st_mode|0o2770) & 0o2777)
+                except PermissionError:
+                    pass
+                for filename in filenames:
+                    filename = os.path.join(dirname, filename)
+                    try:
+                        os.chmod(filename, (os.stat(filename).st_mode|0o660) & 0o777)
+                    except PermissionError:
+                        pass
 
     def start(self):
         if sys.platform == 'win32':
             self.fail("Sorry, the exchange is not available on Windows.")
 
-        self.ensure_root()
+        if not self.coursedir.groupshared:
+            # This just makes sure that directory is o+rwx.  In group shared
+            # case, it is up to admins to ensure that instructors can write
+            # there.
+            self.ensure_root()
```

Comment:
Maybe it would be nice in the groupshared case if the permissions are not set up properly to at least print a warning/error rather than failing with a traceback later (which is what I assume happens if the permissions aren't correct).

Reply:
This checks if the exchange directory is writeable (even when run as a student).  But students only need access to write to {root}/inbound.  So the real issue isn't related to groupshared, it's just that groupshared causes me be a bit more precise, so this got thrown in here.

What do you think we should do?  I can think of at least: a) leave it as is for now, b) make check_exchange_root a configuration option that can be disabled, c) nothing, I make it a separate local change I maintain separately, or d) I allow the root to be writeable by students.

Reply:
I propose (a), it's least invasive and has least risk, and if someone is doing groupshared stuff, they are probably able to debug permission errors pretty easily.

Reply:
Oh oops, I missed this comment before I merged. But yes, I agree with your assessment, I think it's fine to leave it as it is for now.